### DISCUSSION TITLE
Delta Backup

### DISCUSSION THREAD

**Discussion by Jubilee101** [pgmoneta/pgmoneta#146] [Ideas]
## Intro
As I discussed with @jesperpedersen, our next step for pgmoenta would be delta backup. And I'll start working on this today (or tomorrow). The progress will be slower than my gsoc project as I'm still exploring, plus there are a lot of things going on with me right now (God help me 😢). I opened this forum for easier communication.

In my preliminary design, a new `--delta` option will be added to our backup command. And pgmoneta will use the latest backup as the base, after the backup is received from the server, unchanged files will be replaced by symlinks. The comparison process makes use of the file hash value in the manifest file. As a result, delta backup will only work for server version >= 13. If the server version is lower than this, or no backups exist before delta backup, an error will be logged and nothing will be done.

Note that this is just an initial version, since it still has to receive everything before replacing unchanged parts. In the futrue we hope to mitigate this overhead by looking into how to get postgres to only send us the changed parts.

This is a more advanced backup strategy and I can use all the help I can get. Please reach out in this forum if you have suggestions!

## To do
There are roughly 4 parts that needs to be modified:
- [ ] CLI, add a new delta option
- [ ] Backup (of course)
   - [ ] Compare manifest files and replace unchanged parts
   - [ ] Add a persistent label somewhere marking a backup as delta/base backup
   - [ ] Look for ways to have the server only send the changed parts
- [ ] Restore
   - [ ] Replace symlinks with a copy of original files when restoring if identify a backup as a delta backup
- [ ] Retention, we should probably spare the backup serving as base
   - [ ] Discuss strategies here later

**Comment by shikharish** [comment]
What is the `Delta` parameter when I list backups?
```
                   20240327175436 (Backup: 15.8 MB Restore: 83 MB WAL: 32 MB Delta: 0 Retain: No Valid: Yes)
                   20240327175444 (Backup: 9.5 MB Restore: 83 MB WAL: 16 MB Delta: 16 MB Retain: No Valid: Yes)
```

Is this already implemented?

**Reply by jesperpedersen** [reply]
It is the WAL delta - not this feature

**Comment by HazemRawi** [comment]
Will this idea be GSoC'24 project?

**Reply by jesperpedersen** [reply]
No, since it is extremely difficult - but the pgmoneta/Extension project will implement parts of it

**Comment by jesperpedersen** [comment]
@silver-ymz Please, contact us in private first - see the GSoC page

**Comment by silver-ymz** [comment]
I'm sorry for the annoying comments related to gsoc project. I have removed it.
